index.html

```diff
@@ -2497,6 +2502,13 @@ <h3><dfn>Writing content</dfn></h3>
             </li>
           </ol>
         </li>
+        <li>
+          If |context| is `"smart-poster"` and |records| does not contain a
+          <a>URI record</a> for which the protocol is supported by the UA, or
+          if it contains more than one <a>type record</a>, <a>size record</a>
+          or <a>action record</a>, reject |promise| with {{TypeError}} and
```

Comment:
Should we also reject if there are records other than URI record, title record, icon record, size record, type record, action record? e.g. an external type record is embedded.

Reply:
The NFC spec allows that.

> There MAY be other records, which can be treated in an application-specific manner.

Reply:
@zolkis, looks like Web NFC spec doesn't describe this, we should add it to the smart-poster definition.

Reply:
Makes sense, although we cannot aim to reproduce all details of the NFC Forum specs.
Normative text should be in algorithms, and in general it's not a problem they are not duplicated in informative sections.

Reply:
Is a smart-poster record allowed to embed another smart-poster record in the payload? I'm unsure how the NFC spec regulates about that.

Reply:
It doesn't regulate that. If you wonder how the context works with recursive data, well, the algorithms should work AFAICT, even when we embed into an external record a smart poster with an extra smart poster or external which again embed others. The contexts are stacked.

Reply:
@zolkis, does it mean smart-poster allows another mime type? If so, following restriction should be removed.

> If ndef's typeNameField is 2 (MIME type record), then
> If context is equal to "smart-poster" and if running validate smart poster icon type on ndef's type returns false, throw a TypeError and abort these steps.

Reply:
> does it mean smart-poster allows another mime type? 

Yes, it does, but MIME type records have to be representing an icon, see below. 

> A Smart Poster may include an icon by including one or many MIME-
typed image records within the Smart Poster. If the device supports images, it SHOULD
select and display one of these, depending on the device capabilities. The device SHOULD
display only one. The Icon record is optional.

and,

> The Smart Poster may also contain a number of Icon records that have an image NDEF MIME
type [NDEF], [RFC2046]. For example, “image/jpeg”, “image/png”, etc. There might also be
animated Icon records, using common media types such as “video/mpeg”. A reader device
SHOULD display one (and only one) of these Icons to the user prior to acting on the URI record.
A reader device is not required to support any particular image format, but it is
RECOMMENDED that implementers stick to well-known, well-compressed formats such as
PNG [PNG] and JPEG [JPEG].

This pretty much implies that MIME type records in smart poster are _meant_ to represent icons. If it was otherwise, it would have had to be specified explicitly. Therefore I conclude the text should be OK.

Reply:
Got it, thanks for your clarification.

Reply:
By reading the quoted description for defining icon records, it looks to me that MIME records are NOT necessarily acting as icon records.

It just says a Smart Poster can include some MIME records to represent icons, without saying that MIME records must be for icons, i.e. they can be for other application-specific things as well. Otherwise, the spec should have made such a restriction explicitly.

Reply:
So how can you tell if a MIME record was meant as icon or as application-specific thing? I wager _that_ would have had to be explicitly specified by the NFC spec, because it's a confusing situation.

But if MIME type records are considered meant for icons, that is not confusing.

Reply:
If the MIME record's mime type is something like "image/*" or "video/*" that can represent icon data, then it's an icon record. Otherwise, it might be representing some other application-specific data by using various mime types, e.g. JSON data by using "application/json".  If we restrict MIME type to be only for icon data, users lose all other possibilities which seems a surprise and not that friendly.

Now we all agree that the NFC spec allows Smart Poster to include any kind of records,  and it says explicitly that there must have one URI record referring to the smart poster's content and no other url record is allowed, then, when it comes to MIME record, it also should regulate explicitly if there is a similar restriction on that.

Reply:
Well, there can be multiple MIME type records, of which one should be selected by the readers to represent the smart poster icon. So all MIME type records are accepted that qualify for an icon. 

It's true the NFC spec leaves it a bit open what qualifies for an icon. But if we accept any MIME types in smart poster that are _not_ meant to be icons, that will be confusing for readers, since there is no algorithm specified by NFC Forum that would tell how to tell  icons from compatible, but not-meant-for-icon records.

Remember the smart poster use case is meant to describe content available on a network resource (like MIME type, but extended), specifying URL, title, size, type, icon, etc. There is no point having other MIME records than icons, since the content is supposed to be fetched from network and its type is given by the type record.

Apps could also use external records for strictly application purpose media included in smart poster, so there is no limitation

I suggest sticking to the specified algorithm.

Reply:
Well I noticed this note in https://pr-preview.s3.amazonaws.com/zolkis/web-nfc/pull/536.html#dfn-validate-smart-poster-icon-type seems to be saying what I wanted to say :) 

> NOTE
The [NDEF-SMARTPOSTER] specification allows an "image" or "video" MIME type that is supported by applications. Examples are given as "image/jpeg", "image/png", "video/mpeg", but there are no hard requirements on accepted media types. If an application does not find a supported MIME type, it can choose to not display any icon for the smart-poster.

"image/*" and "video/*" are just some examples of mime types to represent icons, not hard requirements. We can allow all mime types and leave applications to choose from them.

Reply:
I meant that `image/jpeg` is an example but expect that all icon media types start with either `image/` or `video/`. Are you arguing for removing that restriction and accept any media types, remove the "validate icon type" steps and move that note?

Reply:
Ah I see what "examples" does mean. And yes I feel that removing this restriction and leaving applications to do whatever they want might bring us (WebNFC) more flexibility, given that NFC forum spec does not restrict that explicitly.

Reply:
I need to run that through with security people. Allowing all MIME types when expecting an icon. Of course it's app responsibility. The NFC Forum is specific, as it allows "**_image_** NDEF MIME" and "**_animated_** Icon records, using common media types such as video/mpeg. 
I am not convinced we should leave it completely free.

Reply:
Well I just feel we do not necessarily need to add a restriction that is not regulated by the NFC forum, i.e. image or video MIME records CAN serve as icon records but other MIME records should still be able to serve for any other purposes.

NFC forum might be specific that only `image` and `video` MIME records are allowed to represent icons, but it is NOT specific on that all MIME records must be for icons.

Security people might have more insights around this.

Reply:
Additional point: from purely use case point of view, other MIME records don't make sense, apps could include those other MIME records outside of the smart-poster. If they want other records inside the smart poster, they could use external or local. So I don't _think_ the NFC Forum _meant_ to allow other MIME records, i.e. I consider it a bug in the NFC Forum spec. Maybe members could ask? @beaufortfrancois ?

Reply:
The threat from security point of view would be this: since apps have to select one MIME record for icon, if the chosen MIME record is not truly compatible with the platform for representing as icon, it might be an attack vector. So restricting icons to certain types would make sense _for the Web_.

Reply:
You mentioned before that NFC forum is saying
> There MAY be other records, which can be treated in an application-specific manner.

I figure this **_other records_** include all record types, not only external or local records?

Reply:
I think restricting the mime type to be only "image/*" or "video/*" does NOT guarantee it must be truly compatible with the platform for icon representation.  The application is supposed to choose an appropriate icon record and even choose whether to display an icon or not.

Reply:
Yes, that is right.

Since this discussion refers writing, we should be extra careful (compared to reading) about what we allow to put there. IMHO it is better to restrict writing of known compatible MIME types for icons (since they figure in the NFC Forum spec).

On the reading side, I think if there is no extra security threat, we could read and expose all records that have been written (supposedly the writing end knew what they were doing). But it would be good to ask the NFC Forum in this matter.

Reply:
Outcome of discussion with security people:
- we should ask NFC Forum how to handle this corner case
- meanwhile, we could have no restrictions on reads (but implementation can harden handling of mime records handling even without the spec requiring it)
- on writes, the problem is that smart poster possibly has associated automatic actions on readers (any kind, not just Web NFC). Therefore allowing any MIME type may result in utilizing a gap in NFC Forum spec or its interpretation among reader implementations. We should try to minimize possible attack surface. So makes sense that Web NFC limits accept smart poster icon mime types to those that are known to be expected to work (image/jpeg, jpg, png etc, video/... etc), if that is not a major restriction for web apps (and should not be).

So I propose that for the moment we restrict on writes only, and we ask/get response from NFC Forum on what is safe to do here.

Reply:
However, we can interpret the NFC Forum spec also in a way that permits extra MIME type records. Note that currently we don't allow external and absolute-url records inside SP.

> There MAY be other records, which can be treated in an application-specific manner. For
example, some application might include a vCard contact card using the proper MIME type.
Applications MAY ignore any extra records inside the Smart Poster.

Reply:
For now removed the restrictions, but we need to track this. It would perhaps be better to limit now and relax later, then restrict later.